### PR TITLE
feat: Enhance PII fine-tuning with multi-GPU support and advanced hyperparameters

### PR DESCRIPTION
**Pull request by wangchen615**
## 🎯 Overview
This PR significantly enhances the PII fine-tuning script with multi-GPU training capabilities, advanced hyperparameter control, and improved model compatibility for production-ready PII detection systems.

### 🔧 **Multi-GPU Training Support**
- Built on Hugging Face Trainer for seamless distributed training
- Compatible with `torchrun` and `accelerate` for multi-GPU setups
- Automatic device detection and optimization

### ⚙️ **Advanced Hyperparameter Control**
- **Learning Rate Scheduling**: Cosine, linear, polynomial, and restart schedulers
- **Optimization**: Configurable learning rate, weight decay, gradient accumulation
- **Regularization**: Dropout control for hidden and attention layers
- **Training Dynamics**: Layer freezing with gradual unfreezing for stability

### 📏 **Enhanced Sequence Length Support**
- Dynamic max sequence length detection from model config
- ModernBERT long context support (up to ~8k tokens)
- Proper tokenization alignment for all model types

### 🧠 **Model Compatibility Improvements**
- Removed MBERT pre-tokenized dependency to prevent misalignment
- Enhanced AutoConfig usage for better model initialization
- Improved ModernBERT compatibility with `reference_compile=False`

### **FreezeLayersCallback**
```python
# Gradual layer unfreezing for improved training stability
FreezeLayersCallback(model, freeze_n_layers=2, unfreeze_after_epochs=1)
```

### **Enhanced Command Line Interface**
- `--learning-rate`: Fine-grained learning rate control
- `--lr-scheduler-type`: Multiple scheduler options
- `--gradient-accumulation-steps`: Memory-efficient training
- `--max-seq-length`: Dynamic sequence length support
- `--freeze-layers` / `--unfreeze-after-epochs`: Training stability control

### **Hugging Face Upload Utility**
- `upload_model.py`: Simple script for model sharing
- Direct integration with Hugging Face Hub

### **High-Performance Training**
```bash
python pii_bert_finetuning.py --mode train --model modernbert-base --dataset ai4privacy \
  --max-seq-length 2048 --batch-size 4 --gradient-accumulation-steps 4 \
  --learning-rate 5e-6 --lr-scheduler-type cosine --dropout 0.3
```

### **Multi-GPU Training**
```bash
torchrun --nproc_per_node=2 pii_bert_finetuning.py --mode train --model modernbert-base
```

### **Layer Freezing Strategy**
```bash
python pii_bert_finetuning.py --mode train --freeze-layers 6 --unfreeze-after-epochs 2
```

## 📊 **Benefits**

| Feature | Before | After |
|---------|--------|-------|
| **GPU Support** | Single GPU only | Multi-GPU with automatic detection |
| **Sequence Length** | Fixed 512 tokens | Dynamic up to 8k (ModernBERT) |
| **Hyperparameters** | Basic | 10+ advanced options |
| **Model Compatibility** | Limited | All BERT variants + ModernBERT |
| **Training Strategy** | Standard | Layer freezing + advanced scheduling |

## 🔧 **Technical Improvements**
- **AutoConfig Integration**: Better model configuration handling
- **Dependency Compatibility**: Fixed NumPy/PyArrow version conflicts
- **Virtual Environment**: Added `pii_env/` to `.gitignore`
- **Code Quality**: Enhanced error handling and logging

## 🧪 **Testing**
- ✅ Tested with ModernBERT, BERT, RoBERTa, DeBERTa
- ✅ Verified multi-GPU compatibility
- ✅ Validated long sequence processing
- ✅ Confirmed model upload functionality

## 📚 **Documentation**
- Updated help text with all new parameters
- Enhanced usage examples in docstring
- Added comprehensive parameter descriptions

## 🔄 **Backward Compatibility**
All existing functionality is preserved. Default parameters maintain original behavior while new features are opt-in.

## 🎯 **Impact**
This enhancement enables production-ready PII fine-tuning with:
- **Faster Training**: Multi-GPU support + gradient accumulation
- **Better Performance**: Advanced hyperparameters + layer freezing
- **Longer Context**: ModernBERT 8k token support
- **Easy Deployment**: Hugging Face integration

### CODE REVIEW

**Comment by rootfs** [comment]
thanks!
